File: content/code/module1/chapter3/node_topic_example.cpp
// Copyright 2025 Open Source Robotics Foundation, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/**
 * Node and Topic Example for Chapter 3: ROS 2 Nodes and Topics.
 *
 * This example demonstrates:
 * - Creating a ROS 2 node in C++
 * - Creating publishers and subscribers
 * - Topic-based communication
 * - Proper node lifecycle management
 */

#include <chrono>
#include <functional>
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"
#include "rclcpp/qos.hpp"

using namespace std::chrono_literals;

/* Publisher Node Class */
class NodeTopicPublisher : public rclcpp::Node
{
public:
  NodeTopicPublisher()
  : Node("node_topic_publisher_cpp")
  {
    // Create a QoS profile for the publisher
    auto qos_profile = rclcpp::QoS(rclcpp::QoSInitialization::with_history_depth(10));
    qos_profile.reliability(RMW_QOS_POLICY_RELIABILITY_RELIABLE);

    // Create a publisher that will publish String messages to the 'chatter_cpp' topic
    publisher_ = this->create_publisher<std_msgs::msg::String>("chatter_cpp", qos_profile);

    // Create a timer that calls the timer_callback method every 500ms
    timer_ = this->create_wall_timer(
      500ms, std::bind(&NodeTopicPublisher::timer_callback, this));

    // Initialize a counter for the messages
    count_ = 0;

    // Log that the publisher node has started
    RCLCPP_INFO(this->get_logger(), "Node Topic Publisher C++ node started");
  }

private:
  void timer_callback()
  {
    // Create a String message
    auto message = std_msgs::msg::String();
    message.data = "Hello, ROS 2 C++! Message #" + std::to_string(count_++);

    // Publish the message
    publisher_->publish(message);

    // Log the message
    RCLCPP_INFO(this->get_logger(), "Publishing: '%s'", message.data.c_str());
  }

  rclcpp::TimerBase::SharedPtr timer_;
  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
  size_t count_;
};

/* Subscriber Node Class */
class NodeTopicSubscriber : public rclcpp::Node
{
public:
  NodeTopicSubscriber()
  : Node("node_topic_subscriber_cpp")
  {
    // Create a QoS profile for the subscriber (should match publisher)
    auto qos_profile = rclcpp::QoS(rclcpp::QoSInitialization::with_history_depth(10));
    qos_profile.reliability(RMW_QOS_POLICY_RELIABILITY_RELIABLE);

    // Create a subscription that will receive String messages from the 'chatter_cpp' topic
    subscription_ = this->create_subscription<std_msgs::msg::String>(
      "chatter_cpp",
      qos_profile,
      std::bind(&NodeTopicSubscriber::topic_callback, this, std::placeholders::_1));

    // Log that the subscriber node has started
    RCLCPP_INFO(this->get_logger(), "Node Topic Subscriber C++ node started");
  }

private:
  void topic_callback(const std_msgs::msg::String::SharedPtr msg) const
  {
    // Log the received message
    RCLCPP_INFO(this->get_logger(), "I heard: '%s'", msg->data.c_str());
  }

  rclcpp::Subscription<std_msgs::msg::String>::SharedPtr subscription_;
};

/* Main function */
int main(int argc, char * argv[])
{
  // Initialize the rclcpp library
  rclcpp::init(argc, argv);

  // Create instances of both publisher and subscriber nodes
  auto publisher_node = std::make_shared<NodeTopicPublisher>();
  auto subscriber_node = std::make_shared<NodeTopicSubscriber>();

  // Use MultiThreadedExecutor to run both nodes simultaneously
  rclcpp::executors::MultiThreadedExecutor executor;
  executor.add_node(publisher_node);
  executor.add_node(subscriber_node);

  // Spin both nodes
  try {
    executor.spin();
  } catch (const std::exception& e) {
    RCLCPP_ERROR(rclcpp::get_logger("main"), "Exception caught: %s", e.what());
  }

  // Clean up: shutdown the rclcpp library
  rclcpp::shutdown();

  return 0;
}

File: content/code/module1/chapter3/node_topic_example.py
#!/usr/bin/env python3
# Copyright 2025 Open Source Robotics Foundation, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
Node and Topic Example for Chapter 3: ROS 2 Nodes and Topics.

This example demonstrates:
- Creating a ROS 2 node
- Creating publishers and subscribers
- Topic-based communication
- Proper node lifecycle management
"""

import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile, ReliabilityPolicy, HistoryPolicy
from std_msgs.msg import String


class NodeTopicPublisher(Node):
    """
    A publisher node that demonstrates creating nodes and publishers.
    """
    def __init__(self):
        # Initialize the parent class (Node) with the node name
        super().__init__('node_topic_publisher')

        # Create a QoS profile for the publisher
        qos_profile = QoSProfile(
            depth=10,
            reliability=ReliabilityPolicy.RELIABLE,
            history=HistoryPolicy.KEEP_LAST
        )

        # Create a publisher that will publish String messages to the 'chatter' topic
        self.publisher_ = self.create_publisher(String, 'chatter', qos_profile)

        # Create a timer that calls the timer_callback method every 0.5 seconds (2 Hz)
        timer_period = 0.5  # seconds
        self.timer = self.create_timer(timer_period, self.timer_callback)

        # Initialize a counter for the messages
        self.i = 0

        # Log that the publisher node has started
        self.get_logger().info('Node Topic Publisher node started')

    def timer_callback(self):
        # Create a String message
        msg = String()
        msg.data = f'Hello, ROS 2! Message #{self.i}'

        # Publish the message
        self.publisher_.publish(msg)

        # Log the message
        self.get_logger().info(f'Publishing: "{msg.data}"')

        # Increment the counter
        self.i += 1


class NodeTopicSubscriber(Node):
    """
    A subscriber node that demonstrates creating nodes and subscribers.
    """
    def __init__(self):
        # Initialize the parent class (Node) with the node name
        super().__init__('node_topic_subscriber')

        # Create a QoS profile for the subscriber (should match publisher)
        qos_profile = QoSProfile(
            depth=10,
            reliability=ReliabilityPolicy.RELIABLE,
            history=HistoryPolicy.KEEP_LAST
        )

        # Create a subscription that will receive String messages from the 'chatter' topic
        self.subscription = self.create_subscription(
            String,
            'chatter',
            self.listener_callback,
            qos_profile)

        # Make sure the subscription is properly created
        self.subscription  # prevent unused variable warning

        # Log that the subscriber node has started
        self.get_logger().info('Node Topic Subscriber node started')

    def listener_callback(self, msg):
        # Log the received message
        self.get_logger().info(f'I heard: "{msg.data}"')


def main(args=None):
    # Initialize the rclpy library
    rclpy.init(args=args)

    # Create instances of both publisher and subscriber nodes
    publisher_node = NodeTopicPublisher()
    subscriber_node = NodeTopicSubscriber()

    # Use MultiThreadedExecutor to run both nodes simultaneously
    try:
        executor = rclpy.executors.MultiThreadedExecutor()
        executor.add_node(publisher_node)
        executor.add_node(subscriber_node)

        # Spin both nodes
        executor.spin()
    except KeyboardInterrupt:
        pass
    finally:
        # Clean up: destroy nodes explicitly
        publisher_node.destroy_node()
        subscriber_node.destroy_node()

        # Shutdown the rclpy library
        rclpy.shutdown()


if __name__ == '__main__':
    main()